src/toil_scripts/lib/urls.py

```diff
@@ -0,0 +1,147 @@
+    """Job version of `download_url`"""
+    work_dir = job.fileStore.getLocalTempDir()
+    fpath = download_url(url, work_dir=work_dir, name=name, s3_encryption_key_path=s3_encryption_key_path)
+    return job.fileStore.writeGlobalFile(fpath)
+
+
+def s3am_upload(fpath, s3_dir, num_cores=1, s3_encryption_key_path=None):
+    """
+    Uploads a file to s3 via S3AM
+    For SSE-C encryption: provide a path to a 32-byte file
+
+    :param str fpath: Path to file to upload
+    :param str s3_dir: Ouptut S3 path. Format: s3://bucket/[directory]
+    :param int num_cores: Number of cores to use for up/download with S3AM
+    :param str s3_encryption_key_path: (OPTIONAL) Path to 32-byte key to be used for SSE-C encryption
+    """
+    assert s3_dir.startswith('s3://'), 'Format of s3_dir (s3://) is incorrect: {}'.format(s3_dir)
+    s3_dir = os.path.join(s3_dir, os.path.basename(fpath))
+    if s3_encryption_key_path:
+        base_url = 'https://s3-us-west-2.amazonaws.com/'
+        url = os.path.join(base_url, s3_dir[5:])
+        temp_key_path = os.path.join(os.path.split(fpath)[0], 'temp.key')
+        with open(temp_key_path, 'wb') as f_out:
+            f_out.write(_generate_unique_key(s3_encryption_key_path, url))
+        _s3am_with_retry(num_cores, '--sse-key-file', temp_key_path,
+                         'file://{}'.format(fpath), s3_dir)
+    else:
+        _s3am_with_retry(num_cores, 'file://{}'.format(fpath), s3_dir)
+
+
+def s3am_upload_job(job, file_id, file_name, s3_dir, num_cores, s3_encryption_key_path=None):
+    """Job version of `s3am_upload`"""
+    work_dir = job.fileStore.getLocalTempDir()
+    fpath = job.fileStore.readGlobalFile(file_id, os.path.join(work_dir, file_name))
+    s3am_upload(fpath=fpath, s3_dir=s3_dir, num_cores=num_cores, s3_encryption_key_path=s3_encryption_key_path)
+
+
+def _download_s3_url(file_path, url):
+    """
+    Downloads from S3 URL via Boto
+
+    :param str file_path: Path to file
+    :param str url: S3 URL
+    """
+    from boto.s3.connection import S3Connection
+    s3 = S3Connection()
+    try:
+        parsed_url = urlparse(url)
+        if not parsed_url.netloc or not parsed_url.path.startswith('/'):
+            raise RuntimeError("An S3 URL must be of the form s3:/BUCKET/ or "
+                               "s3://BUCKET/KEY. '%s' is not." % url)
+        bucket = s3.get_bucket(parsed_url.netloc)
+        key = bucket.get_key(parsed_url.path[1:])
+        key.get_contents_to_filename(file_path)
+    finally:
+        s3.close()
+
+
+def _download_encrypted_file(url, file_path, key_path):
+    """
+    Downloads encrypted files from S3
+
+    :param str url: URL to be downloaded
+    :param str file_path: Output path to file
+    :param str key_path: Path to the 32-byte key file
+    """
+    # Grab master key
+    with open(key_path, 'r') as f:
+        key = f.read()
+    if len(key) != 32:
+        raise RuntimeError('Invalid Key! Must be 32 bytes: {}'.format(key))
```

Comment:
Yeah, I could imagine something though where you wrote 31 bytes of the key into the file, and you're just missing the last byte of the key, or you wrote the key and padded with an extra 0 byte. That'd still be an inadvisable security compromise, IMO.

That being said, I don't feel strongly about this, but given my druthers, I would write out something like `"Bad key in {key_file_path}. Must be 32 bytes!"`.